mlir/lib/Dialect/MQTOpt/Transforms/ToQuantumComputationPattern.cpp

```diff
@@ -321,26 +374,28 @@ struct ToQuantumComputationPattern final : mlir::OpRewritePattern<AllocOp> {
 
     // Update the inputs of all non-mqtopt operations that use mqtopt operations
     // as inputs, as these will be deleted later.
-    // NOLINTNEXTLINE(bugprone-nondeterministic-pointer-iteration-order)
-    for (auto* operation : visited) {
-      if (operation->getDialect()->getNamespace() != DIALECT_NAME_MQTOPT) {
-        updateMQTOptInputs(*operation, rewriter, newAlloc.getQureg(),
-                           measureCount);
-      }
+    for (auto* operation : llvm::make_early_inc_range(mqtUsers)) {
+      mqtUsers.erase(operation); // safe deletion as op will be erased next
+      updateMQTOptInputs(*operation, rewriter, newAlloc.getQureg());
     }
```

Comment:
This seems like a fairly cryptic and LLVM-specific way of doing this (also holds below).
There is a more idiomatic way to do this base don using iterators.
Any `erase` operation from an STL container in C++ will return an updated iterator that can be used for iterating further.
See https://en.cppreference.com/w/cpp/container/vector/erase for an example of how this is done.
Let's rather use that.